feature: add driver for tcn75a temp sensor

danieldegrasse opened the pull request:
Add driver for the TCN75A temperature sensor from microchip. The following features of this sensor are enabled:
- TCN75A oneshot mode, which allows single shot conversions with lower
  power consumtion
- Resolution selection, up to 12 bit resolution (9 bit default)
- Triggering based on temperatue thresholds. If the TCN75A exits a set
  threshold range, the application can be notified via a callback.

The `samples/sensor/thermometer` sample has additionally been updated to support temperature sensors with alert triggers. I have tested the sample on a custom board based on the Kinetis K22, but have provided an overlay for the FRDM K22F for demonstration purposes (I cannot connect the TCN75A to this board to test it)

*Note- this PR is a contribution as part of a personal project, so I apologize if I'm slow to respond to comments*

danieldegrasse committed:
drivers: sensor: introduce driver for TCN75A temperature sensor

Add driver for TCN75A temperature sensor. The following features are
supported:
- TCN75A oneshot mode, which allows single shot conversions with lower
  power consumtion
- Resolution selection, up to 12 bit resolution (9 bit default)
- Triggering based on temperatue thresholds. If the TCN75A exits a set
  threshold range, the application can be notified via a callback.

Signed-off-by: Daniel DeGrasse <daniel@degrasse.com>

bbilas commented on drivers/sensor/tcn75a/tcn75a.c:
```diff
@@ -0,0 +1,132 @@
+	uint8_t adc_conf[2] = {TCN75A_CONFIG_REG, 0x0};
+#endif
+
+	if (!device_is_ready(config->i2c_spec.bus)) {
```
```suggestion
	if (!i2c_is_ready_dt(&config->i2c_spec)) {
```

MaureenHelm commented on drivers/sensor/tcn75a/tcn75a.c:
```diff
@@ -0,0 +1,132 @@
+#endif
+
+	/* Fetch a sample from the 2 byte ambient temperature register */
+	ret = i2c_write_read_dt(&config->i2c_spec, &temp_reg, 1, rx_buf, 2);
```
```suggestion
	ret = i2c_write_read_dt(&config->i2c_spec, &temp_reg, sizeof(temp_reg), rx_buf, sizeof(rx_buf));
```

MaureenHelm commented on drivers/sensor/tcn75a/tcn75a.c:
```diff
@@ -0,0 +1,132 @@
+	static struct tcn75a_data tcn75a_data_##n;                                                 \
+	static const struct tcn75a_config tcn75a_config_##n = {                                    \
+		.i2c_spec = I2C_DT_SPEC_INST_GET(n), TCN75A_TRIGGER(n)};                           \
+	DEVICE_DT_INST_DEFINE(n, &tcn75a_init, NULL, &tcn75a_data_##n, &tcn75a_config_##n,         \
```
```suggestion
	SENSOR_DEVICE_DT_INST_DEFINE(n, &tcn75a_init, NULL, &tcn75a_data_##n, &tcn75a_config_##n,         \
```

MaureenHelm commented on drivers/sensor/tcn75a/tcn75a.h:
```diff
@@ -0,0 +1,83 @@
+#ifdef CONFIG_TCN75A_TRIGGER
+	const struct device *dev;
+	struct gpio_callback gpio_cb;
+	sensor_trigger_handler_t sensor_cb;
```
Need to store a reference to the trigger as well. See #55130

MaureenHelm commented on drivers/sensor/tcn75a/tcn75a_trigger.c:
```diff
@@ -0,0 +1,205 @@
+	/* Save config pointer */
+	data->dev = dev;
+
+	if (!device_is_ready(config->alert_gpios.port)) {
```
```suggestion
	if (!gpio_is_ready_dt(config->alert_gpios)) {
```

MaureenHelm commented on dts/bindings/sensor/microchip,tcn75a.yaml:
```diff
@@ -0,0 +1,17 @@
+
+compatible: "microchip,tcn75a"
+
+include: i2c-device.yaml
```
```suggestion
include: [sensor-device.yaml, i2c-device.yaml]
```

MaureenHelm commented on samples/sensor/tcn75a/prj.conf:
```diff
@@ -0,0 +1,7 @@
+CONFIG_CBPRINTF_FP_SUPPORT=y
+CONFIG_I2C=y
+CONFIG_SENSOR=y
+CONFIG_TCN75A=y
```
CONFIG_I2C and CONFIG_TCN75A shouldn't be needed

MaureenHelm commented on samples/sensor/tcn75a/src/main.c:
```diff
@@ -0,0 +1,125 @@
+#include <stdio.h>
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/gpio.h>
```
Probably not needed

MaureenHelm commented on samples/sensor/tcn75a/src/main.c:
```diff
@@ -0,0 +1,125 @@
+	double current_temp;
+	struct sensor_value low_val;
+	struct sensor_value high_val;
+	const struct sensor_trigger trig = {
+		.chan = SENSOR_CHAN_AMBIENT_TEMP,
+		.type = SENSOR_TRIG_THRESHOLD,
+	};
```
Can you investigate updating `samples/sensor/thermometer/` to conditionally support the threshold trigger, instead of adding a driver-specific sample?

MaureenHelm reviewed (changes requested):
Please add this sensor to `tests/drivers/build_all/sensor`, including the trigger configurations

danieldegrasse committed:
tests: drivers: build_all: add TCN75A to sensor build test

Add TCN75A temperature sensor to sensor build all test, along with
configuration setting to enable driver to use a separate thread for
triggering

Signed-off-by: Daniel DeGrasse <daniel@degrasse.com>

danieldegrasse committed:
samples: sensor: thermometer: add trigger support to sample

Add optional trigger support to sample. If the selected temperature
sensor supports triggering using the AMBIENT_TEMP channel, then
temperature thresholds will be set for +0.5 and +1.5 degrees celsius
above ambient temperature. Sample overlays are provided for the
FRDM_K22F board to demonstrate enabling this feature.

Signed-off-by: Daniel DeGrasse <daniel@degrasse.com>

MaureenHelm commented on tests/drivers/build_all/sensor/sensors_trigger_own.conf:
```diff
@@ -45,3 +45,4 @@ CONFIG_TMP007_TRIGGER_OWN_THREAD=y
 CONFIG_VCNL4040_TRIGGER_OWN_THREAD=y
 CONFIG_WSEN_HIDS_TRIGGER_OWN_THREAD=y
 CONFIG_WSEN_TIDS_TRIGGER_OWN_THREAD=y
+CONFIG_TCN75A_TRIGGER_OWN_THREAD=y
```
Need to do `sensors_trigger_global.conf` and `sensors_trigger_none.conf` too

MaureenHelm replied:
Can you move this line up to preserve alpha order? Same for the other trigger confs

MaureenHelm reviewed (changes requested):
Two small nits, but otherwise looks good.

MaureenHelm commented on tests/drivers/build_all/sensor/i2c.dtsi:
```diff
@@ -728,3 +728,9 @@ test_i2c_lsm6dso16is: lsm6dso16is@6f {
 	irq-gpios = <&test_gpio 0 0>;
 };
+
+test_tcn75a: tcn75a@6a {
+	compatible = "microchip,tcn75a";
+	reg = <0x6a>;
+	alert-gpios = <&test_gpio 0 0>;
+};
```
Update the address to 0x70

danieldegrasse replied:
Looks like another sensor driver merged, so 0x70 is now taken. Address is 0x71 in latest push.

MaureenHelm reviewed (approved):
Rebase to pick up a fix for the failed Pull Request Assigner check

danieldegrasse commented:
@bbilas would you be willing to take another look at this PR?

bbilas commented on drivers/sensor/tcn75a/tcn75a.c:
```diff
@@ -0,0 +1,132 @@
+	/* Set user selected resolution */
+	adc_conf[1] |= TCN75A_CONFIG_RES(config->resolution);
+
+
```
Too many nl here.

bbilas commented on drivers/sensor/tcn75a/tcn75a_trigger.c:
```diff
@@ -0,0 +1,202 @@
+	if (ret < 0) {
+		return ret;
+	}
+	return ret;
```
Missing nl before the return.

bbilas commented on drivers/sensor/tcn75a/CMakeLists.txt:
```diff
@@ -0,0 +1,6 @@
+# Copyright 2022 Daniel DeGrasse <daniel@degrasse.com>
```
Actually it's 2023 :)

danieldegrasse replied:
Code was written in 2022. I've updated copyrights.

bbilas commented on drivers/sensor/tcn75a/tcn75a.c:
```diff
@@ -0,0 +1,132 @@
+	if (ret) {
+		return ret;
+	}
+	data->temp_sample = sys_get_be16(rx_buf);
```
Missing nl before `data->temp_sample`.

bbilas commented:
> @bbilas would you be willing to take another look at this PR?

Sorry for the late reply, but I was traveling. I've reviewed your PR and made some comments.

danieldegrasse commented on drivers/sensor/tcn75a/tcn75a.c:
```diff
@@ -0,0 +1,132 @@
+		adc_conf[1] |= TCN75A_CONFIG_SHUTDOWN;
+	}
+
+#ifdef CONFIG_TCN75A_TRIGGER
```
I've added code to check if the user supplied an ALERT gpio pin, which would only be used with the trigger modes. If no ALERT pin is supplied, then these checks won't be run and the trigger setup for the sensor won't occur.

bbilas replied:
It's still ifdefed under `CONFIG_TCN75A_TRIGGER` Kconfig option so how can it be working with 2 modes simultaneously?

bbilas replied:
@danieldegrasse  Fair enough. Thanks for the clarification :)
